library/std/src/path.rs

```diff
@@ -2038,7 +2038,11 @@ impl Path {
     /// If the path is a normal file, this is the file name. If it's the path of a directory, this
     /// is the directory name.
     ///
-    /// Returns [`None`] if the path terminates in `..`.
+    /// Returns [`None`] if path:
+    /// - Is empty.
+    /// - Terminates in `..` (double dots).
+    /// - Is root (e.g. `/`, `/.`, `/././.`).
+    /// - Only contains current directory (e.g. `.`, `././.`).
```

Comment:
I think this is missing the Prefix variant of Component -- maybe it is more useful to say "returns None if the last component of the path is not Normal" and link to https://doc.rust-lang.org/nightly/std/path/enum.Component.html?

Reply:
That's a great idea, however, here is the documentation of Component:

![image](https://user-images.githubusercontent.com/38900226/130272891-18733ff3-d7d0-4e01-9d15-29f7120a2d77.png)

It does not explain that running components on "a/." gives `Component::Normal("a")` instead of `Normal("a")` AND `CurrrentDirectory`, plus I think people are expecting "asd/.".file_name() to return ".", so, maybe this should be explained somewhere?

Maybe I could add examples to the docs of `Component` itself to show what iterating on `Path::new("/a/b/c")` retrieves, then we could also cover this corner cases.

However, there is already something like this at https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.components.

```rust
use std::path::{Path, Component};
use std::ffi::OsStr;

let mut components = Path::new("/tmp/foo.txt").components();

assert_eq!(components.next(), Some(Component::RootDir));
assert_eq!(components.next(), Some(Component::Normal(OsStr::new("tmp"))));
assert_eq!(components.next(), Some(Component::Normal(OsStr::new("foo.txt"))));
assert_eq!(components.next(), None)
```

So, maybe I should add more examples to `path::components` instead? Or add some explanation in the `Component` page itself.

Waiting for some feedback on these thoughts.

Reply:
I think adding more docs to Component makes sense, since it's a dedicated page for this. We can point there from file_name() and potentially components() as well.

Reply:
Nice, I agree! I've been getting very busy this days, so I might need to delay this PR just a couple weeks, but thanks for the review.